src/player.cpp

```diff
@@ -512,17 +514,17 @@ void player::action_taken()
 void player::update_morale()
 {
     // Decay existing morale entries.
-    for( size_t i = 0; i < morale.size(); i++ ) {
+    for( size_t i = morale.size(); i --> 0; ) {
```

Comment:
>Ahh, the `-->` operator, also known as "goes to". However, it does not exist. Please don't write it as if it were an actual single operator.

If it's against accepted code style, I won't use it, but it's not `goto` and I don't see any harm in this little "semantic sugar" which is absolutely for free. The only drawback IMHO is that it's not so wide spread and without a space can confuse an unfamiliar person.

> If you want to iterate over a container in reverse, use this idiom with iterators instead:

IMO `foo -->` is more obvious than `foo.erase( --bar.base() )`. It's good to write it one time in a template, but not on a regular basis. Also, `erase()` shifts all elements after the deleted many times. `remove_if()` looks more preferable.

>Or alternatively, which may be easiest to understand and read, use two loops.

Why mix functional and imperative approach to looping inside one method? Instead of `for(` it could be:

`std::for_each( morale.begin(), morale.end(), []( morale_point &m ) { m.proceed(); } );`

> Btw. do you even need to process the morale points in reverse?

In your last example (two loops) I certainly don’t. In my case I was iterating backwards because I don't particularly like changing counter inside a loop if it's already changed by the loop itself.

Your pieces of code are smart :) and type-independent, but IMHO a bit cumbersome to write time and again, especially if we know that the type is not going to change. Maybe we could at least write templates for the most common cases?

Thank you for the detailed comment.